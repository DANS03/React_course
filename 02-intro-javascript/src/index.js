console.log('Welcome to 07-Destructuring-Arrays');

const characters = ['Batman' , 'Superman', 'WonderWoman'];

const  [ , ,p3] = characters;

console.log(p3);

const retunrArray = ()  => {
    return ['ABC', 123];
};

const [letters, numbers] = retunrArray();

console.log(letters , numbers);

const state = (value) => {
    return [value , () => { console.log('Hello World') } ];
};

const arr1 = state('Hi');

const [name , greetings] = arr1;

console.log(name);
greetings();