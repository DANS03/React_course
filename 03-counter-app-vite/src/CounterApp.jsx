import PropTypes from 'prop-types';
import React from 'react';
import { useState } from 'react';




export const CounterApp = ({value}) => {


  const [ counter, setCounter ] = useState( value ); 
  
  //console.log('render'); //This is a demostrations that every time that we excecute any of the function the whole component will be render

  function handleClick(){
  
    setCounter( counter + 1);//when we call the setCounter function we are telling react that the state has change there for react must re-render the component

    //setCounter( ( newCounter ) => { newCounter + 1 } )   if we  don't hace access to an element of the hook we can also use this method

  }  

  function handleSustraction(){
  
    setCounter( counter - 1);

  }  
  function handleReset(){
  
    setCounter( value );

  }  




  return (

    <>
    
    <div>CounterApp</div>

    <h2> { counter } </h2>

    <button onClick={ handleClick  }> {/*If you need to pass args to the fucntion you must do it inside the arrow function otherwise you can do it like below */ }
    {/*<button onClick={ handleClick  }> --------Theres no need for parentesis ()*/}
       +1 
    </button>

    <button aria-label="btn-minus" onClick={ handleSustraction  }>{/* aria-label="btn-minus" is an identifire that willbe usa as an id in the test*/}
       -1 
    </button>

    <button onClick={ handleReset }> 
       Reset
    </button>

    </>
    
  )
};


CounterApp.propTypeS = {
    value: PropTypes.number,
};