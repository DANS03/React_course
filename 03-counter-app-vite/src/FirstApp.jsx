import { Fragment } from "react"
import  PropTypes  from "prop-types";// this will enable us tu select the the type of prop that we need


const newMessage="Im a var!"//if you want to print a var in order to be render you must use {}
const objectMessage={
    message:"I'm an object!",
    title:"Object Message!"
};

const getGreetins = (   ) => {
    const name ='John Doe'; //prompt('Plesae enter you name:','John Doe');


    const greeting = "Welcome "+ name +" !  How are you today?" 

    return greeting;
};


//the properties or props is an object that are being send to the component dependig of what is needed
//the most commund way of using it is by destructuring the props
//export const FirstApp = ( props ) => {
export const FirstApp = ( {title , subtitle, name } ) => {
  /* 
    if( !props.title ) { //this is a way of validating the props but it has many downsides that why we use propTypes
        throw new Error('Unexisting title');
    } ;
     */
    return (

    <div id="firstAppDiv"> {/*this div has no reason to exist but is very commont the alternatives is using a fragment*/}
        
        <h2>{ getGreetins() }</h2>

        <div>FirstApp</div>
        
        <p>I'm a subtitle</p>
        
        <p>{ title }</p> 
        <p>{ subtitle }</p>
        <p>{ name }</p>

        <p>{ newMessage }</p> {/* you can use a permited expresion inside of the {} when it's not an object*/}

        <p>{ JSON.stringify(objectMessage) }</p>{/* if you want to use an object then you must convert it to a JSON */}
    
    </div>
 /* 
 <Fragment>
        
    <div>FirstApp</div>
        
    <p>I'm a subtitle</p>
    
</Fragment>
 */

//a synonim to this is :

/* 
 <>
    <div>FirstApp</div>
        
    <p>I'm a subtitle</p>
    
</>
 */
        
  )
};


FirstApp.propTypes = {
    title: PropTypes.string.isRequired, //here you will include the type and which promps you need
    subtitle:PropTypes.number,
};


FirstApp.defaultProps = { //In case that you dont want to define all the props you can use this...
    title:'No title',
    subtitle:'No subtitle' ,
    name: 'David'
};
