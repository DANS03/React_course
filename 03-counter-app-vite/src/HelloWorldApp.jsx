
export function App(){
    //this is an easy way to render a component equivalent to
        //document.createElement.... 
    return <h1>Hello World from REACT!!!</h1>  
};

//export default App;  

//A usefull shortcut is :  rafc     This will create a component using the file name