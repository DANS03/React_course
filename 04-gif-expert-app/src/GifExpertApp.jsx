import React, { useState } from 'react'
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid';

export const GifExpertApp = () => {
    
    const [categories,setCategories ] = useState([ 'Pingu']);

    const addCategories = ( newCategory ) => {

        if( categories.includes(newCategory) ){
            alert('The category already exists');
            return
        }

        setCategories( [ newCategory , ...categories ] );

     };

    return (
    <>

        <h1>Gif Expert App</h1>

        <AddCategory 
        //setCategories = {setCategories} 
        onNewCategory = { (value) => addCategories(value) }//this will be resive inside addCategory and executed there
        />

        { 
            categories.map( (category) => (//the map function executes the code for every item inside the list
            
                <GifGrid category={category} />
            ))
        }
           
    </>
  )
};
