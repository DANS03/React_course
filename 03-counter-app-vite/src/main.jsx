//The extention JSX means that the code is a combination of javascript and html in one file.

import React from "react";
import ReactDOM  from "react-dom/client";
import { App } from "./HelloWorldApp";
import { FirstApp } from "./FirstApp";

//In order to rendender the app we must select the part of the HTML doc where our app will be render, the we activate the strict mode of react so we can have more control over the way that the code is written and inside that we render our component app.

ReactDOM.createRoot( document.getElementById( 'root' ) ).render(                
    <React.StrictMode>
        <App></App>

        <FirstApp></FirstApp>
    </React.StrictMode>
);