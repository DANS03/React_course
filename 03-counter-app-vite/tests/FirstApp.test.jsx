import { getByText, render } from "@testing-library/react"
import { FirstApp } from '../src/FirstApp'


describe('Tests in FirstApp', () => { 
    
    test('Must match the snapshot', () => { 
        
        const greeting = "I'm a subtitle";

        const { container, getByText } = render( <FirstApp  subtitle={1}/>)//The function render does multiple things in this case we want it to return an object with some properties

        expect( container ).toMatchSnapshot();
    });//this test is only usefull when you don't want the rendered page to change


    test('Must find greetitngs ', () => { 

        const { container, getByText } = render( <FirstApp  subtitle={1}/>)//The function render does multiple things in this case we want it to return an object with some properties

        const greeting = "I'm a subtitle";

        expect( getByText(greeting) ).toBeTruthy();

        const h2 = container.querySelector('h2');
         
        expect (h2.innerHTML).toContain('Welcome John Doe !'); 
    });

    test('Get test-id from title', () => { 

        const title = "I'm a test!"

        const { getByTestId } = render(  <FirstApp  title={title} subtitle={1}/> );

        //expect( getByTestId('test-title')).toBeTruthy(); this si an option but is only telling us that the id exists 

        expect( getByTestId('test-title').innerHTML ).toContain(title);

    });

    test('Get name twice', () => { 

        const name = "David" 

        const { getAllByText } = render(  <FirstApp  name={name}/> );//getAllByAltText return a array 


        expect( getAllByText(name).length ).toBe(2);

    });


 });