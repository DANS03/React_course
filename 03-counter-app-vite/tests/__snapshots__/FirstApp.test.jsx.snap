// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in FirstApp Must match the snapshot 1`] = `
<div>
  <div
    id="firstAppDiv"
  >
     
    <h2>
      Welcome John Doe !  How are you today?
    </h2>
    <div>
      FirstApp
    </div>
    <p>
      I'm a subtitle
    </p>
    <p
      data-testid="test-title"
    >
      No title
    </p>
     
    <p>
      1
    </p>
    <p>
      David
    </p>
    <p>
      David
    </p>
    <p>
      Im a var!
    </p>
     
    <p>
      {"message":"I'm an object!","title":"Object Message!"}
    </p>
  </div>
</div>
`;
